sensor_board/common/pinav_parser.c

```diff
@@ -0,0 +1,266 @@
+
+	// Scale to millionths of a degree
+	int32_t millionths_of_degrees = (six_100000s_of_degrees * 10) / 6;
+
+	// check N/S
+	if (it[9] != ',') {	// If this isn't safe, we would have failed by now
+		return FAIL;
+	}
+	if (it[10] == 'N') {
+		*lat = millionths_of_degrees;
+	}
+	else if (it[10] == 'S') {
+		*lat = 0 - millionths_of_degrees;
+	}
+	else {
+		return FAIL;
+	}
+
+	return SUCCESS;
+}
+
+// Private helper function for parsing out longitude from gga sentence
+// uint8_t * it should point to the start of the longitude field
+static generic_status_t parse_gga_long(int32_t * lng, uint8_t * it) {
+	int32_t degrees;
+	if (!stoi((uint32_t *)&degrees, it, 3)) {
+		return FAIL;
+	}
+	int32_t minutes;
+	if (!stoi((uint32_t *)&minutes, it + 3, 2)) {
+		return FAIL;
+	}
+	if (it[5] != '.') {
+		return FAIL;
+	}
+	int32_t ten_thousandths_of_minute;
+	if (!stoi((uint32_t *)&ten_thousandths_of_minute, it + 6, 4)) {
+		return FAIL;
+	}
+
+	//calculate 600,000ths of a degree
+	int32_t six_100000s_of_degrees = degrees * 600000;
+	six_100000s_of_degrees += minutes * 10000;
+	six_100000s_of_degrees += ten_thousandths_of_minute;
+
+	// Scale to millionths of a degree
+	int32_t millionths_of_degrees = (six_100000s_of_degrees * 10) / 6;
+
+	// check N/S
+	if (it[9] != ',') {	// If this isn't safe, we would have failed by now
+		return FAIL;
+	}
+	if (it[10] == 'E') {
+		*lng = millionths_of_degrees;
+	}
+	else if (it[10] == 'W') {
+		*lng = 0 - millionths_of_degrees;
+	}
+	else {
+		return FAIL;
+	}
+
+	return SUCCESS;
+}
+
+// Private helper function to find next comma 
+// Returns number of characters >= 0 between *it and next occurance of ','
+// Returns -1 if ',' not found
+static int8_t next_comma(uint8_t * it) {
+	int8_t count = 0;
+	while (*it != ',') {
+		if ((*it == '\r') || (*it == '\n') || (*it == '\0')) {
+			return -1;
+		}
+		++count;
+	}
+	return count;
+}
+
+
+// Private helper function to parse the valid gga sentence
+// pointed to by sentence into the output struct pointed to by out
+static pinav_parser_status_t parse_gga(pinav_parse_output_t * out, uint8_t * sentence){
+	uint8_t * it; // iterator for parsing sentence
+	int8_t to_next_comma;	// distance to next comma in sentence
+
+	out->id = GGA;
+
+	it = sentence + 7; // place iterator at start of fix time field
+	if (!parse_gga_fix_time(&(out->data.gga.fix_time_millis), it)) {
+		return SENTENCE_FORMAT_ERROR;
+	}
+	if (it[10] != ',') {	// If this isn't safe, parse_gga_fix_time would have failed
+		return SENTENCE_FORMAT_ERROR;
+	}
+	it += 11; // Place iterator at start of lat field 
+	if (!parse_gga_lat(&(out->data.gga.latitude), it)) {
+		return SENTENCE_FORMAT_ERROR;
+	}
+	it += 12; // Place iterator at start of long field
+	if (!parse_gga_long(&(out->data.gga.longitude), it)) {
+		return SENTENCE_FORMAT_ERROR;
+	}
+
+	/*to_next_comma = next_comma(it);
+	if (to_next_comma < 1) {
+		return SENTENCE_FORMAT_ERROR;
+	}
+	else {
+		it += (to_next_comma + 1);
+	}*/
```

Comment:
Commented out/dead code should probably be removed before merge. That's what commit histories are for =P